Adding Cytoscape Documentation to dash-docs (WIP, 3/7)

xhlulu opened the pull request:
## Description

With this PR we'll be introducing Dash-docs. This is a WIP, and please wait for @chriddyp to merge it (even if it's LGTY). 

I've reorganized the commits for clarity. Please look at the commits rather than the individual files changed if you wish to review the code style. For the content, please check out the heroku deployment.

Please reach out to me if you have any extensive question about Cytoscape. I will be at the office more often since semester is only starting.

## Chapter Progression

Here's the progression of the chapters. This is not final, since we might reorder them, add new chapters, remove some, etc. [S] means that it has just been started, [D] indicates that the draft is ready to be reviewed, [C] indicates it is completed.

1. Elements [D]
2. Layout [D]
3. Styling [S]
4. Callbacks [S]
5. User Interaction Events [S]
6. Applications [S]
7. Reference [C]

xhlulu commented:
Big thank you to all the folks at @plotly/dash-bio, and your feedback really helped me for the cytoscape release notes! I have some great feedback from @mkcor still pending that I'll review this week.

chriddyp commented:
🎉 🎉 !

chriddyp commented:
Just took a quick scan through the review app, it's looking great so far!

xhlulu commented:
Thank you!

mkcor commented on tutorial/chapter_index.py:
```diff
@@ -154,6 +156,17 @@
+        'description': '(New! Released TBA) Dash Cytoscape is our new network ' \
+                       'visualization component. It offers a declarative and ' \
+                       'pythonic interface to create beautiful, customizable, ' \
+                       'interactive and reactive graphs'
```
End with a period.

xhlulu replied:
Done

mkcor commented on tutorial/chapter_index.py:
```diff
@@ -605,6 +618,74 @@
+        'content': tutorial.cytoscape.layout_chapter.layout,
+        'name': 'Cytoscape Layouts',
+        'description': '''
+        Description of built-in layouts, and how to modify its properties. 
```
*their* properties?

xhlulu replied:
Thanks for the catch!

xhlulu replied:
Done

mkcor commented on tutorial/chapter_index.py:
```diff
@@ -605,6 +618,74 @@
+        'content': tutorial.cytoscape.elements_chapter.layout,
+        'name': 'Cytoscape Elements',
+        'description': '''
+        An overview of elements declaration and manipulation. 
```
elements’ declaration?

xhlulu replied:
Element declaration would make more sense I think.

mkcor commented on tutorial/chapter_index.py:
```diff
@@ -605,6 +618,74 @@
+        'content': tutorial.cytoscape.reference_chapter.layout,
+        'name': 'Cytoscape Reference',
+        'description': '''
+        A comprehensive list of all of the Cytoscape properties.
```
Sometimes descriptions end with periods and sometimes they don't... Stick with one convention.

xhlulu replied:
Didn't notice that, thanks for pointing out. I'll stick with period (same as datatable)

mkcor commented on tutorial/chapter_index.py:
```diff
@@ -605,6 +618,74 @@
+        'name': 'Cytoscape events',
+        'description': '''
+        Overview of user-interaction events that trigger callbacks in Dash,
+        and how to use it to update itself
```
how to use *them*?
to update *the app*?

xhlulu replied:
Made it clearer, thanks!

mkcor commented on tutorial/chapter_index.py:
```diff
@@ -605,6 +618,74 @@
+        'content': tutorial.cytoscape.styling_chapter.layout,
+        'name': 'Cytoscape Styling',
+        'description': '''
+        Learn how to style elements with a CSS-like syntax
```
Stick with one syntactic form for all descriptions:

* either a substantive (Description, Overview, Examples, ...)
* or a verb (Learn, Style, Combine, ...)

xhlulu replied:
Good idea! I'll keep it consistent with datatable (substansive)

mkcor commented on tutorial/cytoscape/elements_chapter.py:
```diff
@@ -0,0 +1,260 @@
+    
+    Each element is defined by a dictionary declaring its purpose and 
+    describing its properties. Usually, you specify what group the 
+    element belongs to (i.e. if it's a node or an edge), indicate what 
```
American English prefers a comma after "i.e."

xhlulu commented on tutorial/cytoscape/layout_chapter.py:
```diff
@@ -0,0 +1,274 @@
+    '''),
+
+    dcc.Markdown(dedent('''
+    Similarly for the circle layout, we can force the nodes to start and end at
```
Here I meant "Similarly to the grid layout". I'll modify this

xhlulu replied:
Done

xhlulu commented on tutorial/cytoscape/layout_chapter.py:
```diff
@@ -0,0 +1,274 @@
+        style={'width': '100%', 'height': '350px'},
+        layout={
+            'name': 'breadthfirst',
+            'roots': '[id = "nyc"]'
```
This will likely cause an error.

mkcor replied:
Oh, excuse me, nested quotes! Please disregard.

xhlulu commented on tutorial/cytoscape/applications_chapter.py:
```diff
@@ -0,0 +1,29 @@
+import dash_core_components as dcc
+import dash_html_components as html
+import pandas as pd
+from textwrap import dedent
```
thanks for pointing out! big oversight from my part

xhlulu replied:
Done

xhlulu commented on tutorial/cytoscape/callbacks_chapter.py:
```diff
@@ -0,0 +1,31 @@
+import dash_html_components as html
```
Done

xhlulu commented on tutorial/cytoscape/callbacks_chapter.py:
```diff
@@ -0,0 +1,31 @@
+import dash_table
+from .utils import section_title
+from tutorial import tools
+from tutorial import styles
```
Done

xhlulu commented on tutorial/cytoscape/elements_chapter.py:
```diff
@@ -0,0 +1,260 @@
+
+from .utils import CreateDisplay
+from tutorial import tools
+from tutorial import styles
```
Done

xhlulu commented on tutorial/cytoscape/elements_chapter.py:
```diff
@@ -0,0 +1,260 @@
+    - `id`: The index of the element, useful when you want to reference it
+    - `label`: The name associated with the element if you wish to display it 
+    
+    If your element is an edge, the following items are required for your data
```
Pardon if i'm wrong, but isn't a dictionary item the key-value pair? I could say keys if it's less confusing, but here it could be used interchangeably.

xhlulu replied:
Done

xhlulu commented on tutorial/cytoscape/elements_chapter.py:
```diff
@@ -0,0 +1,260 @@
+    - `source`: The `id` of the source node, which is where the edge starts
+    - `target`: The `id` of the target node, where the edge ends
+    
+    The position dictionary takes as items the `x` and `y` position of the node. 
```
Done

xhlulu commented on tutorial/cytoscape/elements_chapter.py:
```diff
@@ -0,0 +1,260 @@
+    `position`.
+    
+    In the case of `data`, the typical keys fed to the dictionaries are:
+    - `id`: The index of the element, useful when you want to reference it
```
This would be possible, but the cytoscape.js uses id as a property, and since element has a 1-1 correpondance we kept id. Since we are using a dictionary declaration, id here is purely used as a string, so shouldn't conflict with the built-in function `id()`. However, this is a good concern that can be addressed in `dash-cytoscape`, and for Dash itself.

xhlulu commented on tutorial/cytoscape/layout_chapter.py:
```diff
@@ -0,0 +1,274 @@
+    '''),
+
+    dcc.Markdown(dedent('''
+    Here is what would happen if we choose Montreal and Vancouver instead:
```
Done

xhlulu commented on tutorial/cytoscape/layout_chapter.py:
```diff
@@ -0,0 +1,274 @@
+
+    dcc.Markdown(dedent('''
+    > Notice here that we are not giving the ID of the nodes to the `roots`
+    > key, but instead use a specific syntax to select the desired elements. 
```
Done

xhlulu commented:
@mkcor Thank you so much for the extensive feedback! I'll keep what you suggested in mind as I go through the other chapters!

xhlulu commented on tutorial/cytoscape/elements_chapter.py:
```diff
@@ -0,0 +1,260 @@
+    dcc.Markdown(dedent('''
+    > Note that those boolean properties can be overwritten by certain Cytoscape
+    > parameters such as `autoungrabify` or `autounselectify`. Please refer to
+    > the reference for more information.
```
Figured out how to do it, done! Thanks

mkcor commented:
@xhlulu Great! Poke me when you're ready for final review and (my) approval. I see there are file conflicts and CI failures; also I know you'll wait for @chriddyp's approval to merge anyway!

xhlulu commented:
Thanks Marianne! I'll start by finishing up the other chapter, modify the current chapters a bit depending on Chris's feedback. Not sure what's the problem with CI, I'll solve it thursday when i'll come by the office. I'll ping you again when the other drafts are finished!

xhlulu commented:
Oh I see, feel free to post them again!

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/50985376-2fd9c780-14d2-11e9-89bc-8050236ddbc9.png)

I'd say "in the official Cytoscape documentation."

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/50985441-50098680-14d2-11e9-97e4-411bba58096b.png)

On the home page, between the chapters and the example, let's add a little block quote informing users that it's interactive. Something like:
> Cytoscape graphs are interactive! Click and drag on the nodes to re-arrange them, scroll to zoom, click and drag on the canvas to drag.

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/50985884-606e3100-14d3-11e9-8dd3-709b394d1781.png)

For all of these titles, let's prefix them with "Cytoscape", e.g. "Cytoscape Layout" or "Cytoscape Stylesheet". That way, if a user lands on this page directly from e.g. google or some other link, they know that they are in the cytoscape docs.

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/50990909-5a7f4c80-14e1-11e9-9d22-b22d16c89d8e.png)

Also, we can remove the double header here and just have "Cytoscape Layout"

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/50991224-2ce6d300-14e2-11e9-9ef4-c617361946a4.png)

Is there any place in the cytoscape docs that we can link to that shows all of the available options?

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/50991414-91a22d80-14e2-11e9-82b1-8b13c58a161a.png)

Could we add a border around these examples? That is, similar to this style where the left border of the code block above the example connects with the full border around the example
![image](https://user-images.githubusercontent.com/1280389/50991467-b39bb000-14e2-11e9-99f4-5d73d140807f.png)

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/50991488-c615e980-14e2-11e9-849e-c08996050e59.png)

Are classes used for anything besides style? If not, I'd just say like "Classes are used to style groups of elements."

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/50992093-383afe00-14e4-11e9-9313-32059616b33d.png)

Could we update the property docstrings to include reference links to the relevant cytoscape docs? i.e. here for stylesheet, then maybe we include a link to http://js.cytoscape.org/#style? similarly with the other properties as well.

I think that folks that visit this "Reference" page will need links to all of these sections, so if we don't include them in the doc strings, maybe we include a bulleted list at the top of the page linking to the relevant parts of the cytoscape docs.

xhlulu commented:
> Is there any place in the cytoscape docs that we can link to that shows all of the available options?

I've included a link to layouts at the beginning of the chapter. However, it does seem that it is not super clear here, so I added a few links and updated the description for clarity.

xhlulu commented:
> Are classes used for anything besides style? If not, I'd just say like "Classes are used to style groups of elements."

For now, it is only useful for styling. However, it might be useful for other selectors actions that are not yet added, e.g. `eles.filter()`, described [here](http://js.cytoscape.org/#eles.filter). I'll change the phrasing for now, and if we ever add other features that uses `classes` I'll update that part.

xhlulu commented:
> ![image](https://user-images.githubusercontent.com/1280389/50992093-383afe00-14e4-11e9-9313-32059616b33d.png)
> 
> Could we update the property docstrings to include reference links to the relevant cytoscape docs? i.e. here for stylesheet, then maybe we include a link to http://js.cytoscape.org/#style? similarly with the other properties as well.
> 
> I think that folks that visit this "Reference" page will need links to all of these sections, so if we don't include them in the doc strings, maybe we include a bulleted list at the top of the page linking to the relevant parts of the cytoscape docs.

This is a good idea. I'm opening up an issue on dash-cytoscape and priotize it for 0.0.4.

xhlulu commented:
@chriddyp Thank you for all the great comments! I've modified the chapters extensively following your comments.

mkcor commented on tutorial/cytoscape/applications_chapter.py:
```diff
@@ -0,0 +1,309 @@
+    coloring and expansion of Cytoscape graphs. Additionally, there are many
+    possible ways to integrate Cytoscape with a library such as biopython,
+    enabling advanced usage in bioinformatics and computational biology. In this
+    chapter, we will show an example of automatically generating phylogeny tree
```
```suggestion
    chapter, we will show an example of automatically generating a phylogenetic tree
```

mkcor commented on tutorial/cytoscape/applications_chapter.py:
```diff
@@ -0,0 +1,309 @@
+    The first step is to load the phylogeny tree. We will be using the 
+    `apaf.xml` file retrieved from [PhyloXML](http://www.phyloxml.org/examples/apaf.xml).
+    
+    To load the file, you simply need to run this (after you made sure that
```
```suggestion
    To load the file, run this (after you made sure that
```

mkcor commented on tutorial/cytoscape/applications_chapter.py:
```diff
@@ -0,0 +1,309 @@
+    `apaf.xml` file retrieved from [PhyloXML](http://www.phyloxml.org/examples/apaf.xml).
+    
+    To load the file, you simply need to run this (after you made sure that
+    biopython is correctly installed):
```
```suggestion
    biopython was correctly installed):
```

mkcor commented on tutorial/cytoscape/applications_chapter.py:
```diff
@@ -0,0 +1,309 @@
+    
+    Since we are assigning a position to the nodes, we have to use the `preset`
+    layout. Additionally, we need to add specific styles in order to make the
+    phylogeny trees to aesthetically match the traditional methods. We define:
```
```suggestion
    phylogeny trees to match aesthetically the traditional methods. We define:
```

mkcor commented on tutorial/cytoscape/applications_chapter.py:
```diff
@@ -0,0 +1,309 @@
+    
+    At this point, we simply need to create the layout of the app, which will
+    be a simple Cytoscape component. We will also add a callback that will 
+    color all the children of an edge red whenever we hover of that edge. This
```
```suggestion
    color all the children of an edge red whenever we hover on that edge. This
```

mkcor commented on tutorial/cytoscape/utils.py:
```diff
@@ -0,0 +1,104 @@
+        for column in df.columns:
+            row.append(
+                html.Td(
+                    df.iloc[i][column],
```
```suggestion
                    df.loc[i, column],
```

Otherwise you are doing chained indexing, which is not a good practice.
Reference: https://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
(I once covered this in an lesson introducing Pandas: https://github.com/mkcor/advanced-pandas)

xhlulu replied:
Thanks for the heads up! This is an excellent idea, however I realized I'm not using this function anywhere in the Cytoscape docs, so I think I could just remove this function. I'll keep this in mind tho!

mkcor commented on tutorial/cytoscape/utils.py:
```diff
@@ -0,0 +1,104 @@
+        dedent(code),
+        language='python',
+        customStyle=styles.code_container
+    )
```
What's your editor? ;)

xhlulu replied:
I'm using Pycharm, which is supposed to tell me this haha. I guess I probably didn't see this one :/

mkcor replied:
[Apparently](https://stackoverflow.com/questions/36043061/automatically-add-newline-on-save-in-pycharm), by default, PyCharm doesn't add a "newline at end of file" whenever you save a file... So it would make sense that you were missing it.

xhlulu replied:
I see... I guess I could auto-format the code, but I don't really like how they enforce the string length :/ I guess I'll have to play around with my Pycharm settings.

mkcor commented on tutorial/cytoscape/utils.py:
```diff
@@ -0,0 +1,104 @@
+                        odd_row_style if i % 2 == 1 else {},
+                        cell_style,
+                        cell_style_by_column.get(column, {}),
+                        conditional_cell_style(df.iloc[i][column], column)
```
```suggestion
                        conditional_cell_style(df.loc[i, column], column)
```

xhlulu replied:
Same as above

mkcor commented on tutorial/examples/cytoscape/usage-phylogeny.py:
```diff
@@ -0,0 +1,198 @@
+        """Create a mapping of each clade to its column position."""
+        depths = tree.depths()
+        # If there are no branch lengths, assume unit branch lengths
+        if not max(depths.values()):
```
```suggestion
        if max(depths.values()) == 0:
```
Actually, is it what you mean?
"Use `is not` operator rather than `not ... is`. While both expressions are functionally identical, the former is more readable and preferred.
Reference: https://www.python.org/dev/peps/pep-0008/#programming-recommendations

xhlulu replied:
Interesting one, I used the code from the Biopython utils and didn't take a close look at this: http://biopython.org/DIST/docs/api/Bio.Phylo._utils-pysrc.html#draw_ascii

xhlulu replied:
Honestly I'm not sure about the type of `depths.values()`, since it could be either an empty list, or a singleton list with value of 0. If it's the former, then `if max(depths.values()) is None` is the best idea, as you suggested. However, if it's the former, then `if max(depths.values()) == 0` should work. Although I believe that the behavior should be either one or the other, it's hard for me to say if the `biopython` library has a consistent behavior for depth evaluation (e.g. if the user accidentally sets the branch lengths of every clades in a tree to be 0, then should we force the branch length to be 1, i.e., `depths(unit_branch_lengths=True)`)? I think this is a good issue, and I'll revisit it as soon as I find a greater collection of phylo data so i can thoroughly test the Phylo object.

mkcor replied:
I see! My hesitation was about `depths.values()` being possibly `None`... but then `max(depths.values())` would give a `TypeError` so we'd have to handle the exception (even with the original syntax). So it looks like the case we're handling here really is `max(depths.values()) == 0`. Should we poke the Biopython folks? ;)

xhlulu replied:
THat's a good idea! I think I can open up an issue there and share you the link

mkcor replied:
Oops, I didn't see your longer comment meanwhile... Well, `max([0]) == 0` is true and `max([])` gives a `ValueError` so, again, it's about exception handling and/or being more defensive upstream: Do we have some `assert` statements to check the existence and type of `depths.values()` in the Cytoscape code?
Testing popular use cases in the Phylo field is also a good idea!

xhlulu replied:
Most of the popular examples contain branch length, which is why I haven't had the chance to extensively test the Phylo object haha

xhlulu replied:
https://github.com/biopython/biopython/issues/1910

xhlulu replied:
^Just added the link to the issue I just created on the biopython repo

mkcor replied:
Great, @xhlulu! By the way, it's easier if you tag me directly in the issue description, so that I get the initial and subsequent (when people reply) notifications ;)

mkcor commented on tutorial/examples/cytoscape/usage-phylogeny.py:
```diff
@@ -0,0 +1,198 @@
+@app.callback(Output('cytoscape-usage-phylogeny', 'stylesheet'),
+              [Input('cytoscape-usage-phylogeny', 'mouseoverEdgeData')])
+def color_children(edgeData):
+    if not edgeData:
```
```suggestion
    if edgeData is not None:
```
Actually, is it what you mean? There might be some type conversion going on...
Reference: https://www.python.org/dev/peps/pep-0008/#programming-recommendations

xhlulu replied:
Will update in next commit

mkcor commented on tutorial/examples/cytoscape/event_callbacks_3.py:
```diff
@@ -0,0 +1,87 @@
+@app.callback(Output('cytoscape-selectedNodeData-markdown', 'children'),
+              [Input('cytoscape-event-callbacks', 'selectedNodeData')])
+def displaySelectedNodeData(data_list):
+    if not data_list:
```
```suggestion
    if data_list == []:
```
Again, we want to be explicit...

xhlulu replied:
Yes, I'll update it to `if data_list is None` (since Dash initialize with `None`)

xhlulu commented on tutorial/chapter_index.py:
```diff
@@ -731,6 +743,74 @@
+        'content': tutorial.cytoscape.callbacks_chapter.layout,
+        'name': 'Cytoscape Callbacks',
+        'description': '''
+        Methods to combine Dash Callbacks to update your Cytoscape object.
```
Updated!

xhlulu commented:
Thanks Marianne for those great reviews! Is there any section you feel needs to be further reviewed? I'll try to have the docs ready by the end of next week, so that we can have it publish on Tuesday.

chriddyp commented:
Looking good! Some small comments below:

***

Let's make this a link to the github page, users won't know what this is referring to. Could even add some more context like "check out the "[`usage-stylesheet.py`](...) example in the dash-cytoscape GitHub project.".
![image](https://user-images.githubusercontent.com/1280389/51858884-c9dba580-2303-11e9-9f1f-97f35e09a1e0.png)

***

Missing a syntax highlighter here

![image](https://user-images.githubusercontent.com/1280389/51858969-0dceaa80-2304-11e9-8469-c2861ecbb07a.png)

***
![image](https://user-images.githubusercontent.com/1280389/51859023-28a11f00-2304-11e9-8250-5d1ee0dd179a.png)

We usually say "client side"

***

![image](https://user-images.githubusercontent.com/1280389/51859042-35257780-2304-11e9-850c-ccf6878d1223.png)

Remove the "(e.g. on DDS)" and say "if this is hosted on a server"

***
> Note that if this simple app is hosted (e.g. on DDS), this callback will behave correctly for all users, since the callback only depend on the current elements of the user's graph (which is stored on the user side), which we input as a State (which doesn't trigger a callback, but is observed when the callback is triggered by an Input). This method is preferred over directly modifying the elements list stored on the server because it restricts any data update to the user side.

I'd simplify this to be something like:

> It's important to _mutate_ the `elements` object by passing it into the callbacks as `State` rather than making it a `global` variable. In general, `global` variables should be avoided as they won't work when multiple users are viewing the app or when the app is deployed with multiple gunicorn workers.

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/51859238-a9601b00-2304-11e9-9352-2f2f89b966c7.png)

Dash users aren't necessarily familar with JSON, so I'd say:
> Elements in Cytoscape are designed to be clear, simple, and declarative.

***

![image](https://user-images.githubusercontent.com/1280389/51859297-cd236100-2304-11e9-97e8-3b0ed4406a91.png)

I think I originally suggested this, however I can't seem to get this to work anymore. Does this work for you?

***

![image](https://user-images.githubusercontent.com/1280389/51859521-520e7a80-2305-11e9-960a-8ad5db466809.png)

I think this is too JS specific, most dash users won't know what `cy.on` is referring to. 

I would simplify this with something like:

```
[Dash callbacks](/getting-started-part 2) allow you to update your Cytoscape graph via other components like dropdowns, buttons, and sliders.
```

***

![image](https://user-images.githubusercontent.com/1280389/51859783-fa244380-2305-11e9-8726-22368eaeb760.png)

Let's change this to "Dash Cytoscape Reference" since we frequently reference the JS cytoscape docs.

***
In the example, you're actually dumping it into a `html.Pre`

![image](https://user-images.githubusercontent.com/1280389/51859842-1b852f80-2306-11e9-9e6c-3193cfec7b86.png)

***
![image](https://user-images.githubusercontent.com/1280389/51859996-846ca780-2306-11e9-86b5-493fd3530758.png)
Can we display this example in the docs?

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/51860037-a23a0c80-2306-11e9-966c-6a6dd6d0a36a.png)

This makes it sound like we did this example in the previous section. I might consider ⚡️ the first three sentances and start with "In this chapter, we will show an example of automatically generating a "

***

![image](https://user-images.githubusercontent.com/1280389/51860193-0361e000-2307-11e9-8c2e-10b997108f98.png)

Similarly link to the actual file on github. You might want to do a find-and-replace to find all of the instances of `usage-*.py`

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/51860227-1bd1fa80-2307-11e9-84aa-85fd6b94f98f.png)

Since this chapter only has a phylogeny tree / biopython example, let's rename it from "Cytoscape Applications" to "Cytoscape with BioPython". When we add new applications from different packages (e.g. networkx), we can create new chapters for them.

chriddyp commented:
![image](https://user-images.githubusercontent.com/1280389/51860325-550a6a80-2307-11e9-9e15-cf7f8073d385.png)

Let's say:
```
The development for this component was sponsored by one of our commercial partners. Interested in steering the roadmap? [Get in touch](https://plot.ly/products/consulting-and-oem/).
```

xhlulu commented:
> Can we display this example in the docs?

I will create a link to the gallery instead of embedding the example in the code, as to reduce memory usage and reduce need for updates if `usage-stylesheet.py` changes.

xhlulu commented:
@cldougl Please take a look at the deployment metrics for this app: https://dashboard.heroku.com/apps/dash-docs-pr-385/metrics/web?ending=0-hours-ago&starting=24-hours-ago

By reducing the number of workers to 4, we decrease the RAM usage by half roughly. Is there any reason why we are using 8 workers on the deployment?

chriddyp commented:
>  Is there any reason why we are using 8 workers on the deployment?

Yes. Performance. The Dash User Guide is a highly traffic-ed website.

What we could do is make this an ENV variable so that there are different settings on review apps vs master (review apps could use 2 workers and the master app could use 8 workers).

xhlulu commented:
@chriddyp I added all the changes you suggested! Let me know if there's anything else to do before merging (on tuesday) :)

xhlulu commented:
Seems like the tests are failing again, but the heroku app is fine!

xhlulu commented:
@chriddyp @cldougl any idea why the tests are failing again? thanks!
